### PR TITLE
fix: exec command with failFast should fail immediately

### PR DESCRIPTION
**Pull request by MohiuddinM**
## Description

The expected behavior of failFast option is to fail the whole exec command as soon as a single fail occurs, and it works as expected if concurrency=1. But if concurrency > 1, it diverges from the expected behavior and instead keeps running until all pooled jobs have finished executing.

This PR fixes this divergence, and the exec command will fail immediately on a single fail irrespective of concurrency value.

## Type of Change

- [ ] ✨ `feat` -- New feature (non-breaking change which adds functionality)
- [x] 🛠️ `fix` -- Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ `!` -- Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 `refactor` -- Code refactor
- [ ] ✅ `ci` -- Build configuration change
- [ ] 📝 `docs` -- Documentation
- [ ] 🗑️ `chore` -- Chore

### CODE REVIEW

**Comment by MohiuddinM** [comment]
Note to maintainers: only the last commit is related, others are showing because I didn't fork specifically for this fix, and instead merged this into my fork.

File: packages/melos/lib/src/commands/exec.dart [hunk 1]
```diff
@@ -117,52 +119,65 @@ mixin _ExecMixin on _Melos {
       packages.map((package) => MapEntry(package.name, Completer<int?>())),
     );
 
-    await pool.forEach<Package, void>(sortedPackages, (package) async {
-      if (failFast && failures.isNotEmpty) {
-        return;
-      }
+    CancelableOperation<void>? operation;
```
Comment (spydon) [inline packages/melos/lib/src/commands/exec.dart:122]:
```suggestion
    late final CancelableOperation<void> operation;
```

File: packages/melos/lib/src/commands/exec.dart [hunk 2]
```diff
@@ -117,52 +119,65 @@ mixin _ExecMixin on _Melos {
...
         }
-      }
 
-      if (!prefixLogs) {
-        logger
-          ..horizontalLine()
-          ..log(AnsiStyles.bgBlack.bold.italic('${package.name}:'));
-      }
+        if (packageExitCode > 0 && failFast) {
+          await operation?.cancel();
```
Comment (spydon) [inline packages/melos/lib/src/commands/exec.dart:174]:
```suggestion
          await operation.cancel();
```

**Review by spydon — COMMENTED** [review]
Thanks for your contribution, the change makes.
Could you add a test, so that we can avoid regression in the future?

**Comment by MohiuddinM** [comment]
@spydon I have added:
- Additional logger to report canceled tasks separately from failed jobs:
```
            '$ melos exec\n'
            '  └> exit 1\n'
            '     └> FAILED (in 1 packages)\n'
            '        └> c (with exit code 1)\n'
            '     └> CANCELED (in 2 packages)\n'
            '        └> a (due to failFast)\n'
            '        └> b (due to failFast)\n'
```
- Added tests
- Refactored earlier implementation to remove exit (to make code testable)

**Comment by MohiuddinM** [comment]
Tests are flaky I guess because all 3 are being run at the same time, and some times "a" fails first, and "b", "c" are cancelled. Otherwise, some other fails. I need to find a way to add delay before "exit 1" so it becomes more predictable.

**Comment by MohiuddinM** [comment]
@spydon I found a solution to the random failure order by adding delays (see: https://github.com/invertase/melos/pull/665/commits/1ab3177051b704a199f7db8b1e86e91acc0c9c08).

The test failing now is unrelated I don't know why, but it passes on my PC.

**Review by spydon — APPROVED** [review]
Lgtm

**Comment by mdikcinar** [comment]
Hi, It looks like this PR broke failFast.

Expected Behavior (Version 5.2.0)
When a test fails, execution stops immediately due to failFast. Which works well on 5.2.0.

Current Behavior (Version 6.2.0)
Even though failFast is enabled, tests continue running instead of stopping at the first failure.

My output on 5.2.0
```melos run test:select
  └> melos exec --fail-fast -c 1 -- "very_good test -- --fail-fast"
     └> RUNNING

$ melos exec
  └> very_good test -- --fail-fast
     └> RUNNING (in 25 packages)

-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
package1:
Running "flutter test" in . ...
✓ Optimizing tests (0.1s)
00:08 +9: All tests passed!
package1: SUCCESS
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
package2:
Running "flutter test" in . ...
✓ Optimizing tests (0.1s)
00:01 +2: All tests passed!
package2: SUCCESS
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
package3:
Running "flutter test" in . ...
✓ Optimizing tests (0.1s)
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════

════════════════════════════════════════════════════════════════════════════════════════════════════

Test failed. See exception logs above.
The test description was: ***

example_test.dart (FAILED)

00:02 +8 -1: Some tests failed.
Failing Tests:
 - example_test.dart 
        - [ERROR] ***

-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

$ melos exec
  └> very_good test -- --fail-fast
     └> FAILED (in 1 packages)
        └> package3 (with exit code 69)

melos run test:select
  └> melos exec --fail-fast -c 1 -- "very_good test -- --fail-fast"
     └> FAILED
ScriptException: The script test:select failed to execute.

melos run test
  └> melos run test:select --no-select
     └> FAILED
```

My output on 6.2.0

```melos run test:select
  └> melos exec --fail-fast -c 1 -- "very_good test -- --fail-fast"
     └> RUNNING

$ melos exec
  └> very_good test -- --fail-fast
     └> RUNNING (in 25 packages)

-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
package1:
Running "flutter test" in . ...
✓ Optimizing tests (0.1s)
00:08 +9: All tests passed!
package1: SUCCESS
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
package2:
Running "flutter test" in . ...
✓ Optimizing tests (0.1s)
00:01 +2: All tests passed!
package2: SUCCESS
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
package3:
Running "flutter test" in . ...
✓ Optimizing tests (0.1s)
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════

════════════════════════════════════════════════════════════════════════════════════════════════════

Test failed. See exception logs above.
The test description was: ***

example_test.dart (FAILED)

00:02 +8 -1: Some tests failed.
Failing Tests:
 - example_test.dart 
        - [ERROR] ***

-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

$ melos exec
  └> very_good test -- --fail-fast
     └> FAILED (in 1 packages)
        └> package3 (with exit code 69)
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
package4:
     └> CANCELED (in 22 packages)
        └> package4 (due to failFast)
        └> package5 (due to failFast)
        └> package6 (due to failFast)
        └> package7 (due to failFast)
        └> package8 (due to failFast)
        └> package9 (due to failFast)
        └> package10 (due to failFast)
        └> package11 (due to failFast)
        └> package12 (due to failFast)
        └> package13 (due to failFast)
        └> package14 (due to failFast)
        └> package15 (due to failFast)
        └> package16 (due to failFast)
        └> package17 (due to failFast)
        └> package18 (due to failFast)
        └> package19 (due to failFast)
        └> package20 (due to failFast)
        └> package21 (due to failFast)
        └> package22 (due to failFast)
        └> package23 (due to failFast)
        └> package24 (due to failFast)
        └> package25 (due to failFast)
Running "flutter test" in . ...
✓ Optimizing tests (0.1s)
00:04 +11: All tests passed!
package4: SUCCESS
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
package5:
Running "flutter test" in . ...
✓ Optimizing tests (0.1s)
00:11 +449 ~1: All tests passed!
package5: SUCCESS
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
package6:
Running "flutter test" in . ...
✓ Optimizing tests (0.1s)
00:01 +1: All tests passed!
package6: SUCCESS
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
package7:
Running "flutter test" in . ...
✓ Optimizing tests (0.1s)
00:02 +60: All tests passed!
package7: SUCCESS
```

**Pull request merged**